docs/source/csdl-spec.md

```diff
@@ -0,0 +1,630 @@
+  id: String! 
+  name: String!
+  tripId: String!
+}
+
+extend type Rocket @key(fields: "id") {
+  id: String! @external
+  astronaut: Astronaut!
+}
+```
+
+### CSDL: Rockets & Astronauts
+
+This example is non-normative, and represents just one possible way to compose [rockets](#rockets) and [astronauts](#astronauts) into CSDL.
+
+<a name=ex-rockets-and-astronauts href=#ex-rockets-and-astronauts class=listing>Example CSDL composing `rockets` and `astronauts`</a>
+```graphql
+schema @csdl(version: "^1") {
+  query: Query
+}
+
+enum csdl_Graph {
+  ROCKETS    @csdl_endpoint(url: "https://rockets.api.com"),
+  ASTRONAUTS @csdl_endpoint(url: "https://astronauts.api.com"),
+}
+
+type Query {
+  rockets: [Rocket]!       @csdl_resolve(graph: ROCKETS)
+  astronauts: [Astronaut]! @csdl_resolve(graph: ASTRONAUTS)
+}
+
+type Astronaut
+  @csdl_key(graph: ASTRONAUTS, repr: "{ id }")
+  @csdl_key(graph: ROCKETS, repr: "{ id }")
+{
+  id: String!     @csdl_resolve(graph: ASTRONAUTS)
+  name: String!   @csdl_resolve(graph: ASTRONAUTS)
+  tripId: String! @csdl_resolve(graph: ASTRONAUTS)
+  rocket: Rocket! @csdl_resolve(graph: ASTRONAUTS)
+}
+
+type Rocket 
+  @csdl_key(graph: ASTRONAUTS, repr: "{ id }")
+  @csdl_key(graph: ROCKETS, repr: "{ id }")
+{
+  id: String!           @csdl_resolve(graph: ROCKETS)
+  name: String!         @csdl_resolve(graph: ROCKETS)
+  astronaut: Astronaut! @csdl_resolve(graph: ASTRONAUTS)
+  captain: Astronaut    @csdl_resolve(graph: ASTRONAUTS,
+                                      provides: "{ tripId }")
+}
+```
+
+The meaning of these directives is explored the [Directives](#directives) section.
+
+## Composition Pipeline Roles
+
+```mermaid
+flowchart TB
+    subgraph A [subgraph A]
+      schemaA([schema A])
+      endpointA([endpoint A])
+    end    
+    subgraph B [subgraph B]
+      schemaB([schema B])
+      endpointB([endpoint B])
+    end
+    subgraph C [subgraph C]
+      schemaC([schema C])
+      endpointC([endpoint C])
+    end
+    subgraph Producer
+      Composer
+    end
+    subgraph csdl
+      CSDL
+    end
+    subgraph Consumer
+      Router
+    end    
+    A-->Composer
+    B-->Composer
+    C-->Composer
+    Composer-->CSDL
+    CSDL-->Router
+    Router-->Clients
+    Clients-->Router
+```
+
+- **Producer** the bit that generates CSDL. spec will place requirements on composers.
+- **Consumer** the bit that consumes CSDL. spec will place requirements on consumers.
+- **Composer** a kind of producer which composes subgraph schemas into CSDL. (no requirements on consumers, diagrammed for clarity.)
+- **Subgraphs** subgraphs which are composed by the composer (no requirements on subgraphs, diagrammed for clarity.)
+- **Router** a kind of consumer which serves CSDL as a graphql endpoint. no requirements on routers, these are described in the federation spec.
+
+## Structure
+
+A CSDL document must be a valid [GraphQL schema definition language](https://spec.graphql.org/draft/#sec-Type-System) document.
```

Comment:
Does the GraphQL specification actually specify "schema definition language"? I was pretty sure that it was a term we made up to mean "the part of the GraphQL specification that people use in their schemas"

Reply:
i guess we could say "type system definition language," which is what the spec actually calls it.

Reply:
changed it to "graphql interface description language"